ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/State.hs

```diff
@@ -206,11 +207,7 @@ modifyOpenState ImmutableDBEnv { _dbHasFS = hasFS :: HasFS m h, .. } action = do
           -- In case of an unexpected error (not an exception), close the DB
           -- for safety
           ExitCaseSuccess (Left (UnexpectedError {})) ->
-              -- This function ('modifyOpenState') can be called from all
-              -- threads, not necessarily the one that opened the ImmutableDB
-              -- and the registry, so we must use 'unsafeReleaseAll' instead of
-              -- 'releaseAll'.
-              (DbClosed, unsafeReleaseAll _dbRegistry)
+              (DbClosed, cleanUp hasFS st)
```

Comment:
Let `open` return an `OpenState` and `close` take one, so that we can make sure that we only clean up an `OpenState`. I.e., we cannot have accidentally closed the state already before trying to clean it up.

Same for `withOpenState`.